### ISSUE TITLE
gcc can fail to find header files (and libraries?) when building packages

### ISSUE THREAD

**Issue by wizeman** [NixOS/nixpkgs#8676]
The following bash scripts have a function called `badPath`:
```
nixpkgs/pkgs/build-support/cc-wrapper/utils.sh
nixpkgs/pkgs/build-support/gcc-cross-wrapper/utils.sh
nixpkgs/pkgs/build-support/gcc-wrapper-old/utils.sh
```

This function is called from the `gcc` wrapper script, for example, when building a package (but only when `NIX_ENFORCE_PURITY` is `1`). It is used to filter out include paths and library paths passed to gcc, so that only paths in the nix store, in `/tmp` or in `$NIX_BUILD_TOP` are accepted (the remaining are filtered out).

However, if a command such as `gcc -I//tmp//...` is used during compilation, then the provided include path is filtered out even though it shouldn't, because `//tmp//` is equivalent to `/tmp/`, but the function is doing a comparison to see if the path literally starts with `/tmp`. I assume the same happens to the other paths (`$NIX_BUILD_TOP`, `$NIX_STORE` and `/dev/null`).

The approached used in that function is quite brittle, because a path such as `/tmp/../dev/sda` would be accepted even though it should be filtered out (I know this is shouldn't be a security issue, but still... the function is not really doing what it's supposed to).

I believe this function should either be removed, or there should be some kind of normalization done before doing the path comparison. For example, `realpath(1)` could be used, although I don't know whether it would have other implications (e.g. should symlinks be followed? Would that cause breakage?).

(BTW, as an aside, I'm not sure why there are various practically identical copies of this function in nixpkgs...)

Note that this bug actually caused me to waste a significant amount of time debugging an (apparently quite perplexing) build failure when upgrading a package...

**Comment by edolstra** [comment]
Strictly speaking, according to POSIX, paths starting with two slashes are not equivalent with one slash. So build scripts that use `//tmp` should be probably be fixed anyway.

**Comment by wizeman** [comment]
Right, paths starting with 2 slashes are interpreted in an implementation-defined manner. However, paths starting with 3 or more slashes *need* to be interpreted as if they only had 1 slash. So we should fix this.

Also, if the build script uses a`/../` component in the path, it can cause the `badPath` function to exhibit a false positive or a false negative behavior. 

It is especially perplexing when you know the file existed at the time gcc failed, it works fine when you compile manually (even when you try it with `nix-shell` without `--pure`), but still it fails with `nix-build`...

**Comment by vcunat** [comment]
Do we have some cases already that fail because of this? (discounting those specifically constructed to fail)

**Comment by wizeman** [comment]
@vcunat See https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/rustc/generic.nix#L139

Even though I worked around the problem, it may very well reappear in the same package or in different packages. 
But most importantly, I found this extremely hard to debug and it caused me to waste a lot of time, which was very frustrating... I would expect other maintainers to have similar problems (unless they're already aware of this bug) .

**Comment by vcunat** [comment]
I see no reason not to fix this. I expect it to be relatively easy, with using some `realpath` or such...

**Comment by heydojo** [comment]
I could be wrong but it appears that this bug is the reason that I had to patch ati-drivers using patchelf to manually set the path to libstdc++ in some of the binaries.

**Comment by Profpatsch** [comment]
(triage) there’s been a few patches, what’s the status @rasendubi

**Comment by rasendubi** [comment]
@Profpatsch I've tried to fix the issue in the easiest way. However, it turned out, it was not working well, so I abandoned the patch. (I don't know any other way to fix the issue.)

**Comment by wizeman** [comment]
It seems like it's still worth fixing this...

Based on @peti's comment on #17132 it seems that there are people with setups where `/nix` is a symlink.

It seems to me that if `/nix/store` can be a symlink to something else then it might not be a good idea to convert the path using anything that actually follows symlinks, because paths in the Nix store will actually resolve to something whose path doesn't start with `/nix/store`, so there would be no easy way to actually determine if they belong to the Nix store or not.

@peti How about using `realpath -m -q -s` instead of just `realpath -q`?

This means that symlinks and directories would not actually be followed, as far as I understand it would just be normalizing the path based only on its text, e.g. converting `///tmp/////non_existing_dir///../file` to `/tmp/file`.

There is still an issue where the normalized path according to `realpath -m -q -s` can point to something that is not actually the real path that would be accessed. For example, the result of `realpath -m -q -s /tmp/symlink/../file` is `/tmp/file` while the actually path that would be accessed can be something completely different like `/var/somedir/file` (depending on where `symlink` points to), but this wouldn't be any worse than it currently is, and arguably it would be more correct!

I think if we used `realpath -m -q -s` it would solve the issues that I brought up, especially its non-intuitiveness and time-wasting potential. What do you think?
